core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -78,15 +78,19 @@ trait Endpoint[A] { self =>
   /**
    * Maps this endpoint to the given function `A => Future[B]`.
    */
-  final def mapAsync[B](fn: A => Future[B]): Endpoint[B] = new Endpoint[B] {
-    final def apply(input: Input): Endpoint.Result[B] = self(input) match {
-      case EndpointResult.Matched(rem, out) =>
-        EndpointResult.Matched(rem, out.flatMapF(oa => oa.traverse(a => fn(a))))
-      case _ => EndpointResult.Skipped
-    }
+  final def mapAsync[B](fn: A => Future[B]): Endpoint[B] =
+    new Endpoint[B] with (Output[A] => Future[Output[B]]) {
```

Comment:
So we can pass the same instance (`this` in that case) as a parameter to the flatMap/map functions. We don't have to allocate a closure each time. Apparently, mixing a new interface (`Function1` in that case) and then defining a new method is super cheap.